criu/files-reg.c

```diff
@@ -1380,12 +1388,318 @@ static bool should_check_size(int flags)
 	return true;
 }
 
+/* Gets the build-id (If it exists) from 32-bit ELF files.
+ * Returns the number of bytes of the build-id if it could
+ * be obtained, else -1.
+ */
+static int get_build_id_32(Elf32_Ehdr *file_header, unsigned char **build_id,
+				const int fd, size_t mapped_size)
+{
+	int size, num_iterations;
+	size_t file_header_end;
+	Elf32_Phdr *program_header, *program_header_end;
+	Elf32_Nhdr *note_header, *note_header_end;
+
+	file_header_end = (size_t) file_header + mapped_size;
+	if (file_header >= (Elf32_Ehdr *) (file_header_end - sizeof(Elf32_Ehdr)))
+		goto err;
+
+	/* 
+	 * If the file doesn't have atleast 1 program header entry, it definitely can't
+	 * have a build-id.
+	 */
+	if (!file_header->e_phnum) {
+		pr_warn("Couldn't find any program headers for file with fd %d\n", fd);
+		goto err;
+	}
+
+	program_header = (Elf32_Phdr *) (file_header->e_phoff + (char *) file_header);
+	if (program_header <= (Elf32_Phdr *) file_header)
+		goto err;
+
+	program_header_end = (Elf32_Phdr *) (file_header_end - sizeof(Elf32_Phdr));
+	num_iterations = file_header->e_phnum + 1;
+
+	/* 
+	 * If the file has a build-id, it will be in the PT_NOTE program header 
+	 * entry AKA the note sections.
+	 */
+	while (num_iterations-- && program_header <= program_header_end &&
+			program_header->p_type != PT_NOTE)
+		program_header++;
+	
+	if (!num_iterations || program_header >= program_header_end) {
+		pr_warn("Couldn't find the note program header for file with fd %d\n", fd);
+		goto err;
+	}
+
+	note_header = (Elf32_Nhdr *) (program_header->p_offset + (char *) file_header);
+	if (note_header <= (Elf32_Nhdr *) file_header)
+		goto err;
+
+	note_header_end = (Elf32_Nhdr *) (file_header_end - sizeof(Elf32_Nhdr));
+	/* Arbitrary large number to prevent unnecessarily searching the entire file */
+	num_iterations = 500;
+
+	/* The note type for the build-id is NT_GNU_BUILD_ID. */
+	while (num_iterations-- && note_header <= note_header_end &&
+			note_header->n_type != NT_GNU_BUILD_ID)
```

Comment:
```
note_header_end = (Elf32_Nhdr *)(program_header->p_filesz + note_header);
while (note_header <= note_header_end && ...)
```
hmm?

Reply:
Doing this seems to cause job 7 to fail (https://travis-ci.org/github/AnorexicAtticusFinch/criu/jobs/708335971).